cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+
+```
+void harness(void) {
+  poly a;
+  uint8_t r[MLKEM_POLYBYTES];
+
+  /* Contracts for this function are in poly.h */
+  poly_tobytes(r, &a);
+}
+```
+
+### Top-level contracts
+
+The comments on `poly_tobytes()` give us a clear hint:
+
+```
+ * Arguments:   INPUT:
+ *              - a: const pointer to input polynomial,
+ *                with each coefficient in the range [0,1,..,Q-1]
+ *              OUTPUT
+ *              - r: pointer to output byte array
+ *                   (of MLKEM_POLYBYTES bytes)
+```
+So we need to write a REQUIRES contract to constrain the ranges of the coefficients denoted by the parameter `a`. There is no constraint on the output byte array, other than it must be the right length, which is given by the function prototype.
+
+We can use the macros in `cbmc.h` to help, thus:
+
+```
+void poly_tobytes(uint8_t r[MLKEM_POLYBYTES], const poly *a)
+     // clang-format off
+REQUIRES(a != NULL && IS_FRESH(a, sizeof(poly)))
+REQUIRES(ARRAY_IN_BOUNDS(int, k, 0, (MLKEM_N - 1), a->coeffs, 0, (MLKEM_Q - 1)))
+ASSIGNS(OBJECT_WHOLE(r));
+// clang-format on
+```
+
+### Interior contracts and loop invariants
+
+`poly_tobytes` has a reasonable simple, single loop statement:
+
+```
+  for (unsigned int i = 0; i < MLKEM_N / 2; i++)
+  { ... }
+```
+
+A candidate loop contract needs to state that:
+1. The loop body assigns to variable `i` and the whole object pointed to by `r`.
+2. Loop counter variable `i` is in range 0 .. MLKEM_N / 2 at the point of the loop invariant (remember the pattern above).
+3. The loop terminates because the expression `MLKEM_N / 2 - i` decreases on every iteration.
+
+Therefore, we add:
+
+```
+  for (unsigned int i = 0; i < MLKEM_N / 2; i++)
+      // clang-format off
+  ASSIGNS(i, OBJECT_WHOLE(r))
+  INVARIANT(i >= 0)
+  INVARIANT(i <= MLKEM_N / 2)
+  DECREASES(MLKEM_N / 2 - i)
+    // clang-format on
+    { ... }
+```
+
+Another small set of changes is required to make CBMC happy with the loop body. By default, CBMC is pedantic and warns about conversions that potential truncate values of lose information via an implicit type conversion.
+
+In the original version of the function, we have 3 lines, the first of which is:
+```
+r[3 * i + 0] = (t0 >> 0);
+```
+which has an implicit conversion from `uint16_t` to `uint8_t`. This is well-defined in C, but CBMC issues a warning just in case. To make CBMC happy, we have to explicitly reduce the range of t0 with a bitwise mask, and use an explicit conversion, thus:
+```
+r[3 * i + 0] = (uint8_t)(t0 & 0xFF);
+```
+and so on for the other two statements in the loop body.
+
+### Prove it!
+
+With those changes, CBMC completes the proof:
+
+```
+cd cbmc/proofs/poly_tobytes
+make result
+cat logs/result.txt
+```
+concludes
+```
+** 0 of 228 failed (1 iterations)
+VERIFICATION SUCCESSFUL
+```
+
+We can also use the higher-level Python script to prove just that one function:
+
+```
+cd cbmc/proofs
+./run-cbmc-proofs.py --summarize -j8 -p poly_tobytes
```

Comment:
```suggestion
MLKEM_K=3 ./run-cbmc-proofs.py --summarize -j$(nproc) -p poly_tobytes
```
MLKEM_K does not have a default value, so this will fail.

Reply:
updated as suggested.